### ISSUE TITLE
Make DebugLog more SteamClient-personal

### ISSUE THREAD

**Issue by JustArchi** [SteamRE/SteamKit#503]
As part of https://github.com/SteamRE/SteamKit/issues/260#issuecomment-352566952 I noticed that `DebugLog` is always static and prints information about any amount of clients taking part in the process.

Sometimes this might be confusing. An example situation would be when we're debugging one particular `SteamClient` and other `SteamClients` working in the background are generating at best useless, and at worst misleading messages that are not related to our main `SteamClient` in any way. Of course you could say "then just use one SteamClient while reproducing the problem", but it's not always possible or a proper solution, since sometimes bugs are triggered exactly when more than one `SteamClient` is operating.

Therefore my suggestion is to somehow include bot identifier in output messages, so we could make use of it for identifying to which `SteamClient` given message should be connected to.

There are several possibilities to achieve this, the way that would be most native for my ASF would be separating general "static" `DebugLog` from `SteamClient`-specific messages that are now in "another" `DebugLog` specific to given `SteamClient`. This way we can be subscribed to any amount of debug log generators, mainly debug logs of specific bots, as well as existing "static" one.

If that's too much work (e.g. making `SteamClient`-specific `DebugLog` is unwanted), we could at least add and make use of an identifier in `SteamClient` and print it in the message. This could be a simple `string` in `SteamClient` object. This way we could map problem of bots ourselves.

```
2017-12-18 21:27:13|ArchiSteamFarm-27083|INFO|nebrassy|Start() Starting...
2017-12-18 21:27:13|ArchiSteamFarm-27083|DEBUG|ASF|WriteLine() CMClient | nebrassy | Initialized new SteamClient with ID nebrassy
2017-12-18 21:27:14|ArchiSteamFarm-27083|INFO|Pyro|Start() Starting...
2017-12-18 21:27:14|ArchiSteamFarm-27083|DEBUG|ASF|WriteLine() CMClient | Pyro | Initialized new SteamClient with ID Pyro
2017-12-18 21:27:14|ArchiSteamFarm-27083|INFO|Scout|Start() Starting...
2017-12-18 21:27:14|ArchiSteamFarm-27083|DEBUG|ASF|WriteLine() CMClient | Scout | Initialized new SteamClient with ID Scout
```

Of course those are only two suggestions, feel free to suggest other ones.

**Comment by yaakov-h** [comment]
Making it client-specific requires some kind of unique identifier - or separate logger objects - which would be shared through to connection objects, web API, UFS and so forth.

My initial thought is to shove a `DebugLog` instance into `SteamConfiguration`, but we'd probably have to thread that through to a lot more places.

Dependency injection could make it easier, but @xPaw would probably kill me if I added more dependencies to SK2 😄

**Comment by yaakov-h** [comment]
I was thinking of just an integer counter per SteamClient instance, it's easier to read than a GUID. Most people just skip the middle bytes of a GUID anyway when reading one. 😅 

The biggest problem I'm trying to figure out is API surface - do we basically redesign `DebugLog` from scratch? How do we pass that GUID around? Should it be attached to SteamConfiguration, or should we just use something like `AsyncLocal<T>` to keep a handle on the instance/ID? (If so, does it need to be re-entrant?)

**Comment by JustArchi** [comment]
The only issue I can see with `int` is an application that intentionally creates (and perhaps even disposes) a lot of `SteamClient`s across its runtime. Sure, it's not like it's going to hit 2 billion anytime soon, but it's something to keep in mind. We can always reset counter back to `0`, but then we could have some crazy person who would keep first steam client, dispose everyone else and finally have two with the same ID after counter reset (or overflow) 😅. Then again, it's a matter of limit, and GUIDs at some point are guaranteed to deplete too, although one would need a looooot more time for that, and we don't need to go through hoops of verifying that given ID is truly unused by every instance (I already have `Dictionary<int, SteamClient>` in mind, 😅).

API surface is a good question. If you like my idea of having one static ID for everything not a client, and unique ID for everything connected with particular client, then I guess that `DebugLog.WriteLine()` could make use of extra `int id` in addition to `category` and `message`, so consumer could do what he needs with it (print it together with the message or whatever). ID obviously should be immutable and fixed at `SteamClient` object creation (just to keep everybody sane), I'm thinking whether we should use `SteamConfiguration` for that or not, but I'd probably refrain from that as it could possibly allow the same ID to be assigned to several clients, even unintentionally. Of course there is a situation where one could want several clients to share the same ID, but then I'd rather tell him "group IDs in some kind of dictionary yourself" which is far more safe.

My current idea is just to add `ID = GenerateID()` in `SteamClient` constructor, make it `public readonly whatever ID` (or `{ get; private set; }`) and pass around wherever it's needed, it's `readonly` anyway. `GenerateID()` function obviously generates ID in whatever way we decide upon, whether it's `Guid.NewGuid()` or some other `++GlobalID` for `int`.

In fact, now that I think of it we could probably even hook it up to `GetHashCode()` and re-use that one as ID, just either write it in the log or expose in `WriteLine()`. Like I said, a lot of possibilities, and you'll like some of them more than the others 🙂.

**Comment by yaakov-h** [comment]
> I'm thinking whether we should use SteamConfiguration for that or not, but I'd probably refrain from that as it could possibly allow the same ID to be assigned to several clients, even unintentionally.

If we go this route, each SteamClient would have to take a slightly mutated copy of the configuration... which would mean more allocations :(

> In fact, now that I think of it we could probably even hook it up to GetHashCode() and re-use that one as ID

It should work, but it's not guaranteed:

> Although the RuntimeHelpers.GetHashCode method returns identical hash codes for identical object references, you should not use this method to test for object identity, because this hash code does not uniquely identify an object reference. - [MSDN](https://msdn.microsoft.com/en-us/library/11tbk3h9(v=vs.110).aspx)

**Comment by JustArchi** [comment]
> It should work, but it's not guaranteed

I had some evil plan of getting a memory address at some point... 😅

**[ObjectIDGenerator](https://msdn.microsoft.com/en-us/library/system.runtime.serialization.objectidgenerator(v=vs.110).aspx)** looks nice for this purpose. It's probably better than pure `int` while still having more focused usage than rather generic and broad `Guid`.

**Comment by yaakov-h** [comment]
That seems like it would keep a reference to the SteamClient around forever.

Maybe we _should_ just use a GUID.

**Comment by JustArchi** [comment]
> If we go this route, each SteamClient would have to take a slightly mutated copy of the configuration... which would mean more allocations :(

Maybe you could use `Lazy<Guid>` so it gets initialized the moment `SteamClient` asks for it? Or remove from config entirely and use as a property of `SteamClient` that gets initialized in its constructor.

**Comment by yaakov-h** [comment]
I don't think we need to make it lazy. Deferring creation until we actually log a message seems like overkill.

**Comment by yaakov-h** [comment]
Another idea - what if we just give each `SteamClient` an ID, and include that in the `category`?

e.g. `DebugLog.WriteLine( "123456/CMClient", "something happened in CMClient." );` ?

**Comment by JustArchi** [comment]
It's fine as long as you can actually map those IDs to SteamClients later on, so if that will be a property of `SteamClient` then nothing stops consumer from mapping on his own, being fine solution.

Nobody said that it needs to be configurable, only that it should be there 🙂

**Comment by JustArchi** [comment]
On the other hand it could make more sense to leave those as strings, so the consumer can decide how exactly he wants to map those and what to use as identifiers. For example, in my ASF case I'd map SteamClients to appropriate bot names and know right away from which bot the message comes from, saving myself extra noise in the log in form of "Bot ASDF got assigned to SteamClient 1234 ID" and then reading `1234/CMClient something happened".

**Comment by yaakov-h** [comment]
Implemented by #777.
